File: contracts/risk.py
"""The app contract: what the backend sends the SATYACHECK app.

The app renders states, never thresholds. It maps RiskLevel to overlay visuals and
never computes a band from score; DEFAULT_BAND_MAPPING is the agreed fallback and
deployments override it in backend policy config, not in app code.

Continuous scoring: RiskUpdate is sent roughly once a second for the life of the
call, never a single decision at the start.
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Annotated, Literal, Union

from pydantic import BaseModel, Field

from .checks import CheckName, ReasonCode

SCHEMA_VERSION = "1.0"


class RiskVerdict(str, Enum):
    GENUINE = "genuine"
    SYNTHETIC = "synthetic"
    UNKNOWN = "unknown"  # insufficient audio, or a model degraded


class RiskLevel(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


# score -> first RiskLevel at or above it. App renders RiskLevel only.
DEFAULT_BAND_MAPPING: dict[int, RiskLevel] = {
    0: RiskLevel.LOW,
    40: RiskLevel.MEDIUM,
    70: RiskLevel.HIGH,
    90: RiskLevel.CRITICAL,
}


class SessionStart(BaseModel):
    """Call started. App shows a subtle scanning indicator."""

    kind: Literal["session_start"] = "session_start"
    stream_id: str
    call_id: str
    started_at: datetime
    protected_number: str | None = None  # the enrolled user's own number
    schema_version: str = SCHEMA_VERSION


class RiskUpdate(BaseModel):
    """One scoring tick, roughly once per second while the call is live."""

    kind: Literal["risk_update"] = "risk_update"
    stream_id: str
    call_id: str
    sequence: int
    timestamp: datetime
    score: int = Field(ge=0, le=100)
    verdict: RiskVerdict
    risk_level: RiskLevel
    confidence: float = Field(ge=0.0, le=1.0)  # aggregate signal confidence, not a threshold
    reasons: list[ReasonCode] = Field(default_factory=list)
    contributing_checks: list[CheckName] = Field(default_factory=list)
    degraded_checks: list[CheckName] = Field(default_factory=list)
    # Opaque ids only. No audio, no transcript, no numbers.
    evidence_refs: list[str] = Field(default_factory=list)
    schema_version: str = SCHEMA_VERSION


class CallEnded(BaseModel):
    """Terminal message. Carries the stage 07 evidence summary, nothing sensitive."""

    kind: Literal["call_ended"] = "call_ended"
    stream_id: str
    call_id: str
    ended_at: datetime
    duration_seconds: float = 0.0
    final_score: int = Field(ge=0, le=100)
    final_verdict: RiskVerdict
    final_level: RiskLevel
    reasons: list[ReasonCode] = Field(default_factory=list)
    alert_fingerprint: str | None = None  # present iff an alert was raised
    merkle_root: str | None = None  # published off-ledger anchor
    sealed_record_id: str | None = None
    root_published_at: datetime | None = None
    schema_version: str = SCHEMA_VERSION


AppMessage = Annotated[
    Union[SessionStart, RiskUpdate, CallEnded],
    Field(discriminator="kind"),
]

File: web/dashboard-a/src/types/risk.ts
/**
 * Hand-written TypeScript mirror of contracts/risk.py and contracts/checks.py,
 * for reference only. Never imported by contracts/, and not the source of
 * truth: contracts/*.py is. Keep in sync by hand if the Pydantic models change.
 */

export type RiskVerdict = 'genuine' | 'synthetic' | 'unknown'
export type RiskLevel = 'low' | 'medium' | 'high' | 'critical'

export type CheckName = 'machine_fingerprint' | 'speaker_identity' | 'prosody' | 'stt_llm'

export type ReasonCode =
  | 'fingerprint_synthetic'
  | 'fingerprint_genuine'
  | 'voiceprint_no_enrolment'
  | 'voiceprint_no_match'
  | 'voiceprint_match'
  | 'voiceprint_match_synthetic'
  | 'prosody_anomaly'
  | 'prosody_normal'
  | 'script_risk_high'
  | 'script_risk_absent'
  | 'context_high_risk'
  | 'context_low_risk'
  | 'insufficient_audio'
  | 'degraded_check'

export interface SessionStart {
  kind: 'session_start'
  stream_id: string
  call_id: string
  started_at: string
  protected_number: string | null
  schema_version: string
}

export interface RiskUpdate {
  kind: 'risk_update'
  stream_id: string
  call_id: string
  sequence: number
  timestamp: string
  score: number
  verdict: RiskVerdict
  risk_level: RiskLevel
  confidence: number
  reasons: ReasonCode[]
  contributing_checks: CheckName[]
  degraded_checks: CheckName[]
  evidence_refs: string[]
  schema_version: string
}

export interface CallEnded {
  kind: 'call_ended'
  stream_id: string
  call_id: string
  ended_at: string
  duration_seconds: number
  final_score: number
  final_verdict: RiskVerdict
  final_level: RiskLevel
  reasons: ReasonCode[]
  alert_fingerprint: string | null
  merkle_root: string | null
  sealed_record_id: string | null
  root_published_at: string | null
  schema_version: string
}

export type AppMessage = SessionStart | RiskUpdate | CallEnded
